### PR TITLE
StaticFileMiddleware now passes over to the next middleware in the pi…

### PR DESCRIPTION
**Pull request by brentnewbury**
…peline when invalid characters are encountered. Fixes issue #84

### CODE REVIEW

File: test/Microsoft.AspNet.StaticFiles.Tests/StaticFileMiddlewareTests.cs [hunk 1]
```diff
@@ -31,6 +31,18 @@ public class StaticFileMiddlewareTests
             Assert.Equal(HttpStatusCode.NotFound, response.StatusCode);
         }
 
+        [Fact]
+        public async Task InvalidCharacters_PassesThrough()
+        {
+            TestServer server = StaticFilesTestServer.Create(app => app.UseStaticFiles(new StaticFileOptions()
+            {
+                RequestPath = new PathString(""),
+                FileProvider = new PhysicalFileProvider(Path.Combine(Directory.GetCurrentDirectory(), "."))
```
Comment (davidfowl) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileMiddlewareTests.cs]: Whats the point of this:

`Path.Combine(Directory.GetCurrentDirectory(), ".")`

Why not just do `Directory.GetCurrentDirectory()`

Comment (brentnewbury) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileMiddlewareTests.cs]: I thought I'd seen it in another test in the repo, so I was trying to be consistent.

Comment (brentnewbury) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileMiddlewareTests.cs]: It is a piece of silly code though. I was thinking "someone could turn this into a theory, add inline data, and replace the '.'". In the end it just looks like I'm jumping through hoops for nothing.

Would you like me to change it now or ket it ride through?

Comment (davidfowl) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileMiddlewareTests.cs]: Change it. Whoever adds theories can change it back (unless you were planning to add more cases)

**Comment by brentnewbury** [comment]
Hmm, I'm guessing on Linux the < is valid which causes my unit test to fail. Should I replace the < character with Path.GetInvalidPathCharacters()[0]?

What do you recommend @davidfowl?

File: test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs [hunk 2]
```diff
@@ -49,13 +51,54 @@ public void LookupFileInfo_ReturnsTrue_IfFileExists()
...
+            // Arrange
+            var options = new StaticFileOptions();
+            options.FileProvider = new TestFileProvider();
+            var pathString = new PathString("/test");
+            var httpContext = new DefaultHttpContext();
+            httpContext.Request.Path = new PathString($"/test/{invalidPath}.txt");
+            var context = new StaticFileContext(httpContext, options, PathString.Empty, NullLogger.Instance);
+
+            // Act
+            context.MatchPath();
```
Comment (davidfowl) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs]: Why does this need to be called?

File: test/Microsoft.AspNet.StaticFiles.Tests/StaticFileMiddlewareTests.cs [hunk 3]
```diff
@@ -31,6 +31,33 @@ public class StaticFileMiddlewareTests
...
+                app.Run(ctx =>
+                {
+                    ctx.Response.StatusCode = 200;
+                    staticFilesMiddlewarePassedThrough = true;
+                    return Task.FromResult(0);
+                });
+            });
+            HttpResponseMessage response = await server.CreateRequest($"/{invalidPath}.txt").GetAsync();
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            Assert.Equal(true, staticFilesMiddlewarePassedThrough);
```
Comment (davidfowl) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileMiddlewareTests.cs]: `Assert.True`

File: test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs (hunk 2 above)
Comment (brentnewbury) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs]: The call to MatchPath() is required to populate the _subPath as part of the StaticFilesContext. But the entire test could go, it's pretty much already covered as part of the middleware test. I just thought I'd have add a unit test for the changes to StaticFilesContext as well as the integration test with the middleware and starting up a test server.

File: test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs [hunk 4]
```diff
@@ -49,13 +51,54 @@ public void LookupFileInfo_ReturnsTrue_IfFileExists()
...
+        [Fact]
+        public void ValidatePath_ReturnsFalse_IfPathContainsInvalidCharacters()
+        {
+            // Test whole set of invalid chars in one test as testing indivually may be slow due to number of invalid path chars on Windows
+            var invalidPath = new string(Path.GetInvalidPathChars());
+
+            // Arrange
+            var options = new StaticFileOptions();
+            options.FileProvider = new TestFileProvider();
+            var pathString = new PathString("/test");
```
Comment (davidfowl) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs]: Is this used?

Comment (brentnewbury) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs]: Nope, you're right, didn't spot that.

File: test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs [hunk 5]
```diff
@@ -49,13 +51,53 @@ public void LookupFileInfo_ReturnsTrue_IfFileExists()
...
+            Assert.False(validateResult);
+        }
+
+        [Fact]
+        public void ValidatePath_ReturnsTrue_IfPathDoesNotContainInvalidCharacters()
+        {
+            // Arrange
+            var options = new StaticFileOptions();
+            options.FileProvider = new TestFileProvider();
+            var pathString = new PathString("/test");
```
Comment (davidfowl) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs]: here too :smile:

Comment (brentnewbury) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs]: We'll get there, one day, we'll get there :smile:

**Comment by brentnewbury** [comment]
Let me know when you want me to squash my commits.

File: src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs [hunk 6]
```diff
@@ -80,6 +81,10 @@ static LoggerExtensions()
                 logLevel: LogLevel.Warning,
                 eventId: 13,
                 formatString: "Multiple ranges are not allowed: '{Ranges}'");
+            _logPathInvalid = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Warning,
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs]: This should be verbose like _logPathMismatch.

File: test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs (hunk 2 above)
Comment (Tratcher) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs]: Assert MatchPath == true?

**Comment by Tratcher** [comment]
Just minor comments, then squash it.

File: test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs (hunk 2 above)
Comment (brentnewbury) [inline test/Microsoft.AspNet.StaticFiles.Tests/StaticFileContextTest.cs]: I was trying to be consistent with some of the tests which don't assert on MatchPath, but just noticed some do. I've added the asserts to my unit tests as they can't hurt. Will see if the latest commits pass CI and then squash.

**Comment by brentnewbury** [comment]
@Tratcher Squashed. AppVeyor continually queued for some reason, Travis failed with the message
>Failed to fetch http://security.ubuntu.com/ubuntu/dists/trusty-security/main/i18n/Translation-en  Hash Sum mismatch

**Comment by Tratcher** [comment]
Sorry, I was thinking about this some more and I think we've applied the fix to the wrong layer. StaticFilesMiddleware is abstracted over different kinds of file systems via FileProviders. The implementations we ship are for Physical files and Embedded files. The list of invalid characters is different between these two implementations, and probably for others as well. Checking for invalid characters is the responsibility of those implementations, not this middleware. Hence https://github.com/aspnet/StaticFiles/issues/84.

In this instance the only call to a Path API in this repo is inside of the IContentTypeProvider implementation [FileExtensionContentTypeProvider](https://github.com/aspnet/StaticFiles/blob/dev/src/Microsoft.AspNet.StaticFiles/FileExtensionContentTypeProvider.cs#L429). Since that's where the problem is isolated to, that's where we should apply the fix. Otherwise we risk blocking legitimate scenarios. I think we need an implementation of GetExtension that's agnostic across file providers and doesn't choke on weird characters. It just needs to split on the last dot.

**Comment by brentnewbury** [comment]
Sure, well done catching that. Where should we go from here? I'd like to take a stab at that change if possible.

**Comment by Tratcher** [comment]
Go ahead and try it. You can re-use this PR or create a new one.

**Comment by brentnewbury** [comment]
@Tratcher Wondering what to do with the unit test for this change as the only invalid path character on Unix based systems is the string terminator \x00. So I think if I try to add the following unit test, the Travis build will fail:
```
[Fact]
public void InvalidCharactersAreIgnored()
{
    var provider = new FileExtensionContentTypeProvider();
    string contentType;
    Assert.True(provider.TryGetContentType($"{System.IO.Path.GetInvalidPathChars()}.txt", out contentType));
}
```

Because the string passed in to `TryGetContentType` will appear empty (i.e. \x00.txt).

**Comment by Tratcher** [comment]
`$"file{System.IO.Path.GetInvalidPathChars()}.txt"`? C# allows nulls in strings, it doesn't terminate them.

**Comment by brentnewbury** [comment]
Ah, that's odd. I've had an issue years ago where a null terminator in the middle of an XML file caused the contents to be improperly parsed, that and the inspector doesn't display characters after the null terminator. You learn something new every day :)

I'll close this pull request and start a new one.

**Pull request closed without merging**